### ISSUE TITLE
Factory generation on read-only file system

### ISSUE THREAD

**Issue by chris-brabender** [Smile-SA/elasticsuite#583]
Factories defined in di.xml do not generate unless defined in a __construct method. This causes problems (complete failures) on read only file systems (eg: Magento Cloud). 

### Preconditions
* Magento 2.2
* Install Smile/ElasticSuite extensions
* Deploy to read only file system

Magento Version : 2.2 EE (Cloud)

ElasticSuite Version : 2.3.10

Environment : Production

### Steps to reproduce
1. Install magento 2.2 EE
2. Install Smile/ElasticSuite extensions
3. Execute setup:di:compile in CI environment to generate build
4. Deploy to read only file system based server
5. Attempt to use site

### Expected result
1. Site functions as normal

### Actual result

For example (multiple factories are not generated though): 

{"0":"The specified \"\/app\/generated\/code\/Smile\/ElasticsuiteCore\/Search\/Request\/Query\/BooleanFactory.php.13402\" file could not be written Warning!file_put_contents(\/app\/generated\/code\/Smile\/ElasticsuiteCore\/Search\/Request\/Query\/BooleanFactory.php.13402):

**Comment by romainruaud** [comment]
Hello @chris-brabender 

Many many thanks for submitting this one.

I used to face a similar issue sometimes (Magento was saying the factory was missing, not that he was not able to generate it, but I think this is strongly related) and never found out what was causing it. I always suspected it could be related to the fact these factories are only "declared" in the di.xml and does not exists in __construct calls.

However, as a Magento integrator and hosting provider, we do host many Magento applications on read-only filesystems, so I am wondering why we did not face this issue more often. I have to contact people internally to ensure how is built our complete deployment process. 

On the other hand, I am also working on containerized deployments which I put the system on read-only after the deployment process, so following your instinct, I should have the same issue (factories not generated during compilation, and failure to generate them on the fly after).

I'll keep you in touch,

thank you again,

Regards

**Comment by romainruaud** [comment]
Oh and one more point, we do have websites running on Magento Cloud too, and did not face this issue as far as I know. I will investigate this point too.

**Comment by afoucret** [comment]
Can you please try using branch 2.4.x ?

This release will bring official support for Magento 2.2.x

**Comment by chris-brabender** [comment]
@afoucret is 2.4 production ready or still in development? Do you have an ETA on when it may be officially release?

**Comment by afoucret** [comment]
No release yet for this branch.
You can test it with :

```composer require "smile/elasticsuite:2.4.x-dev@dev"```

We will release very soon version 2.4.0.

**Comment by afoucret** [comment]
Hi @chris-brabender,

Does the problem still occur when using branch 2.4.x ?

**Comment by chris-brabender** [comment]
Hi @afoucret , It appears to be an issue on 2.4.x also unfortunately.

**Comment by afoucret** [comment]
Hi @chris-brabender,

This bug is related to the Magento recommendation : the generation path should not be writable in production mode.

At the same time, DI compilation does not generate factories declared in the di.xml file. Only PHP file fr factories. It is quite annoying since we declared some factory with the following method :

```
<type name="Smile\ElasticsuiteCore\Search\Request\Query\QueryFactory">
        <arguments>
            <argument name="factories" xsi:type="array">
                <item name="boolQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\BooleanFactory</item>
                <item name="filteredQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\FilteredFactory</item>
                <item name="nestedQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\NestedFactory</item>
                <item name="notQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\NotFactory</item>
                <item name="missingQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\MissingFactory</item>
                <item name="termQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\TermFactory</item>
                <item name="termsQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\TermsFactory</item>
                <item name="rangeQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\RangeFactory</item>
                <item name="matchQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\MatchFactory</item>
                <item name="commonQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\CommonFactory</item>
                <item name="multiMatchQuery" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\MultiMatchFactory</item>
                <item name="functionScore" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\Query\FunctionScoreFactory</item>
            </argument>
        </arguments>
    </type>
```

We have to rewrite this part of the code and audit all factories declared as explained bellow.
At the same time we have to audit all calls to the object manager to get sure the same type of bug will not occurs (basically class that will get generated just in time and not processed during the compilation process).

As we want to release 2.4.0 very soon, we plan to fix this in 2.4.1.
2.4.x branch will contains the patch very soon. If you are ok, you can help us and test the patch when we will release it.

BR,

**Comment by romainruaud** [comment]
As @afoucret said, we will fix it. But I have also submitted this as an issue to Magento since it could also occurs to other projects/modules/people and could be considered as a bug of the compilation engine.

Regards

**Comment by afoucret** [comment]
The patch (#612) is merged on the 2.4.x branch.

@chris-brabender : can you test it ?

**Comment by afoucret** [comment]
Version 2.4.0 has been released and contains the PR. The bug is considered as fixed by us. Feel free to reopen if it is not the case.

**Comment by chris-brabender** [comment]
Hi @afoucret ,

I've upgrade to 2.4.0 and am now getting this error:

        Smile\ElasticsuiteCatalog\Helper\CategoryAttribute
                Incompatible argument type: Required type: string. Actual type: \Magento\Catalog\Model\ResourceModel\Category\Attribute\CollectionFactory; File: 
/var/www/html/local.clientsite.com/vendor/smile/elasticsuite/src/module-elasticsuite-catalog/Helper/CategoryAttribute.php

        Smile\ElasticsuiteCatalog\Helper\ProductAttribute
                Incompatible argument type: Required type: string. Actual type: \Magento\Catalog\Model\ResourceModel\Product\Attribute\CollectionFactory; File: 
/var/www/html/local.clientsite.com/vendor/smile/elasticsuite/src/module-elasticsuite-catalog/Helper/ProductAttribute.php

Total Errors Count: 2

                                           
  [Magento\Framework\Validator\Exception]  
  Error during compilation

**Comment by afoucret** [comment]
Hi @chris-brabender : see issue #614 and PR #615 for this one.

Thx for reporting.

**Comment by a-dite** [comment]
Hi @afoucret,

Just had the same issue (main topic of this issue #583) with Magento Cloud Edition 2.1.10. Used this described fix of Magento support: https://support.magento.com/hc/en-us/articles/115002541893-Error-read-only-file-system-when-accessing-the-read-only-directory-var-generation- . So maybe you can fix it also in pre-2.4 (we are using 2.3.11). Thx
